### PR TITLE
ENT-3191: Collect metadata of cloud providers

### PR DESCRIPTION
**Pull request by jirihnidek**
* Added support for collecting metadata for three
  cloud providers:
  - AWS 
  - Azure
  - Google Cloud Platform
* Implementation uses requests Python package
* Type hints are used in code
* Base class of collectors is implemented in: 
    src/rhsmlib/cloud/collector.py
* AWS: Collecting metadata and signature supports both version
  of IMDS
  - Collector tries to use IMDSv1 first, when cached security
    used by IMDSv2 is not found. IMDSv1 is prefered, because
    it requires less communication with metadata server
  - IMDSv1 can be disabled on some systems. In this situation
    token has to be gathered and then it is possible to get 
    metadata and signature
  - Security token is cached for one hour
  - Metadata is simple json document
  - AWS supports three variants of signature. We get rsa2048
    variant of signature
  - Metadata and signature is not cached, because it could
    cause issues, when system is migrated between datacenters
* Azure: collecting of metadata and signature is very simple
  - Several versions of API are supported and old versions
    are obsoleted after three years. It will be necessary to
    update version of API with every minor version of API 
  - Metadata is provided in simple json document
  - Signature uses pkcs7
  - Metadata nor signature is not cached, because it
    could cause issues during migration of VMs
* GCP: Google uses a little bit different approach. It
  provides metadata in JWT token
  - Token is valid for one hour so it is possible to cache
    content of the token
  - It is not simple to decode content of token. It requires
    Python packages that are not available in RHEL rpms.
    Thus it will not be possible to include any data from
    metadata in facts
* Added several unit tests

### CODE REVIEW

File: src/rhsmlib/cloud/providers/aws.py [hunk 1]
```diff
@@ -112,16 +123,301 @@ def is_likely_running_on_cloud(self):
         return probability
 
 
-# Some temporary smoke testing code. You can test this module using:
-# sudo PYTHONPATH=./src:./syspurse/src python3 -m rhsmlib.cloud.providers.aws
```
Comment (wottop) [inline src/rhsmlib/cloud/providers/aws.py:116]: /syspurpose/

File: src/rhsmlib/cloud/providers/azure.py [hunk 2]
```diff
@@ -94,16 +102,130 @@ def is_likely_running_on_cloud(self):
         return probability
 
 
-# Some temporary smoke testing code. You can test this module using:
-# sudo PYTHONPATH=./src:./syspurse/src python3 -m rhsmlib.cloud.providers.azure
```
Comment (wottop) [inline src/rhsmlib/cloud/providers/azure.py:98]: /syspurpose/

**Comment by wottop** [comment]
Result of suggested smoke test:
http://pastebin.test.redhat.com/929683

**Review by wottop — CHANGES_REQUESTED** [review]

File: src/rhsmlib/cloud/providers/aws.py [hunk 3]
```diff
@@ -112,16 +123,301 @@ def is_likely_running_on_cloud(self):
...
+        print(f'>>> debug <<< metadata signature: {signature}')
+
+        metadata_v2 = metadata_collector._get_metadata_from_server_imds_v2()
+        print(f'>>> debug <<< cloud metadata: {metadata_v2}')
+        signature_v2 = metadata_collector._get_signature_from_server_imds_v2()
+        print(f'>>> debug <<< cloud signature: {signature_v2}')
+
+
+# Some temporary smoke testing code. You can test this module using:
+# sudo PYTHONPATH=./src:./syspurse/src python3 -m rhsmlib.cloud.providers.aws
```
Comment (wottop) [inline src/rhsmlib/cloud/providers/aws.py:421]: /syspurpose/

File: src/rhsmlib/cloud/providers/azure.py [hunk 4]
```diff
@@ -94,16 +102,130 @@ def is_likely_running_on_cloud(self):
...
+    if result is True:
+        azure_cloud_collector = AzureCloudCollector()
+        metadata = azure_cloud_collector.get_metadata()
+        signature = azure_cloud_collector.get_signature()
+        print(f'>>> debug <<< metadata: {metadata}')
+        print(f'>>> debug <<< signature: {signature}')
+
+
+# Some temporary smoke testing code. You can test this module using:
+# sudo PYTHONPATH=./src:./syspurse/src python3 -m rhsmlib.cloud.providers.azure
```
Comment (wottop) [inline src/rhsmlib/cloud/providers/azure.py:229]: /syspurpose/

File: src/rhsmlib/cloud/providers/gcp.py [hunk 5]
```diff
@@ -93,16 +101,180 @@ def is_likely_running_on_cloud(self):
...
+        token = gcp_cloud_collector.get_metadata()
+        print(f'>>> debug <<< 1. token: {token}')
+        # 2. using some custom audience
+        gcp_cloud_collector = GCPCloudCollector(audience_url="https://localhost:8443/candlepin")
+        token = gcp_cloud_collector.get_metadata()
+        print(f'>>> debug <<< 2. token: {token}')
+
+
+# Some temporary smoke testing code. You can test this module using:
+# sudo PYTHONPATH=./src:./syspurse/src python3 -m rhsmlib.cloud.providers.gcp
```
Comment (wottop) [inline src/rhsmlib/cloud/providers/gcp.py:278]: /syspurpose/

**Comment by jirihnidek** [comment]
> Result of suggested smoke test:
> http://pastebin.test.redhat.com/929683

Yeah, it is expected result, when you run code on your laptop. The output could be little bit confusing, because I collect system facts first to be able to detect anything and code gathering system facts prints some strange values.

Did you test it on any VM running on AWS/Azure/GCP?

File: src/rhsmlib/cloud/providers/aws.py [hunk 6]
```diff
@@ -112,16 +123,301 @@ def is_likely_running_on_cloud(self):
...
+        """
+
+        if self._token_exists() is False:
+            # First try to get metadata using IMDSv1
+            metadata = self._get_metadata_from_server_imds_v1()
+
+            if metadata is not None:
+                return metadata
+
+        # When it wasn't possible to get metadata using IMDSv1, then try to get metadata using IMDSv2
```
[the remark this answers is not in the document]
Comment (jirihnidek) [reply src/rhsmlib/cloud/providers/aws.py:285]: New version requires three http requests and the IMDSv1 requires only two http requests. Using security token IMHO doesn't provide more security it only adds another http request.

File: src/rhsmlib/cloud/providers/aws.py [hunk 7]
```diff
@@ -112,16 +123,301 @@ def is_likely_running_on_cloud(self):
...
+        Try to get signature from cache file first. When cache file is not available, then try to
+        get signature from server.
+        :return: None
+        """
+        return super(AWSCloudCollector, self).get_signature()
+
+
+def _smoke_tests():
+    """
+    WIP function for smoke testing on AWS cloud
```
[the remark this answers is not in the document]
Comment (jirihnidek) [reply src/rhsmlib/cloud/providers/aws.py:382]: Yeah, I will remove all these `_smoke_test()` functions from `aws.py`, `azure.py` and `gcp.py`, when detecting and collection of metadata will be used by sub-man.

**Comment by cnsnyder** [comment]
The smoke tests worked for me (turned out true only when running on the requisite cloud platform).

**Review by wottop — APPROVED** [review]

**Pull request merged**